### ISSUE TITLE
the code of iter_size type cannot converge

### ISSUE THREAD

**Issue by darkknightzh** [torch/torch7#982]
Hi, 
There are often two types of training method with batchSize, first is the ordinarily batchSize like the code(in the following part) in method 1, second is iter_size type as in method 2 (like iter_size in caffe). But sometimes I want to get the output as much as possible, and use them in criterion, so as to save memory of GPU, I use the code in method3(both model:forward and model:backward use miniBatchSize), but it can not converge, while the accurate is quite low.
when miniBatchSize is the same as batchSize, all of these three methods will converge. But when miniBatchSize is much smaller than batchSize ( like the value in the following code, batchSize = 2000, miniBatchSize = 200), method 1 and method 2 will both converge, but method 3 will not converge. I do not know what’s wrong with the code of method 3. Can someone figure out it? Thank you all.

**the output of method 1 in the first 20 epochs:**
Epoch: 1 Current loss: 1.874033; validation set accu: 0.747900	
Epoch: 2 Current loss: 0.979035; validation set accu: 0.855300	
Epoch: 3 Current loss: 0.583559; validation set accu: 0.878500	
Epoch: 4 Current loss: 0.455771; validation set accu: 0.891300	
Epoch: 5 Current loss: 0.392498; validation set accu: 0.901600	
Epoch: 6 Current loss: 0.354950; validation set accu: 0.907600	
Epoch: 7 Current loss: 0.328135; validation set accu: 0.913600	
Epoch: 8 Current loss: 0.306808; validation set accu: 0.918100	
Epoch: 9 Current loss: 0.290368; validation set accu: 0.922000	
Epoch: 10 Current loss: 0.275467; validation set accu: 0.924600	
Epoch: 11 Current loss: 0.264399; validation set accu: 0.928600	
Epoch: 12 Current loss: 0.252634; validation set accu: 0.930000	
Epoch: 13 Current loss: 0.242605; validation set accu: 0.933000	
Epoch: 14 Current loss: 0.232813; validation set accu: 0.937700	
Epoch: 15 Current loss: 0.224335; validation set accu: 0.938500	
Epoch: 16 Current loss: 0.217700; validation set accu: 0.942100	
Epoch: 17 Current loss: 0.209554; validation set accu: 0.944000	
Epoch: 18 Current loss: 0.202671; validation set accu: 0.947000	
Epoch: 19 Current loss: 0.196189; validation set accu: 0.945700	
Epoch: 20 Current loss: 0.190535; validation set accu: 0.949300	

**the output of method 2 in the first 20 epochs:**
Epoch: 1 Current loss: 2.020510; validation set accu: 0.691500	
Epoch: 2 Current loss: 1.187532; validation set accu: 0.818500	
Epoch: 3 Current loss: 0.665779; validation set accu: 0.862100	
Epoch: 4 Current loss: 0.492206; validation set accu: 0.885200	
Epoch: 5 Current loss: 0.414528; validation set accu: 0.894500	
Epoch: 6 Current loss: 0.369595; validation set accu: 0.905100	
Epoch: 7 Current loss: 0.337944; validation set accu: 0.910900	
Epoch: 8 Current loss: 0.314416; validation set accu: 0.916200	
Epoch: 9 Current loss: 0.295827; validation set accu: 0.919900	
Epoch: 10 Current loss: 0.279754; validation set accu: 0.924600	
Epoch: 11 Current loss: 0.266030; validation set accu: 0.928500	
Epoch: 12 Current loss: 0.254025; validation set accu: 0.932300	
Epoch: 13 Current loss: 0.243666; validation set accu: 0.935300	
Epoch: 14 Current loss: 0.233726; validation set accu: 0.936600	
Epoch: 15 Current loss: 0.225125; validation set accu: 0.940800	
Epoch: 16 Current loss: 0.217339; validation set accu: 0.941100	
Epoch: 17 Current loss: 0.209984; validation set accu: 0.942800	
Epoch: 18 Current loss: 0.203375; validation set accu: 0.946100	
Epoch: 19 Current loss: 0.196839; validation set accu: 0.945500	
Epoch: 20 Current loss: 0.191074; validation set accu: 0.947300	

**and the output of method 3 in the first 20 epochs:**
Epoch: 1 Current loss: 2.383539; validation set accu: 0.136500	
Epoch: 2 Current loss: 2.348151; validation set accu: 0.135500	
Epoch: 3 Current loss: 2.327148; validation set accu: 0.134700	
Epoch: 4 Current loss: 2.312341; validation set accu: 0.138400	
Epoch: 5 Current loss: 2.301245; validation set accu: 0.143100	
Epoch: 6 Current loss: 2.292610; validation set accu: 0.147700	
Epoch: 7 Current loss: 2.285001; validation set accu: 0.156300	
Epoch: 8 Current loss: 2.278466; validation set accu: 0.166600	
Epoch: 9 Current loss: 2.271928; validation set accu: 0.175200	
Epoch: 10 Current loss: 2.266094; validation set accu: 0.185600	
Epoch: 11 Current loss: 2.260496; validation set accu: 0.195000	
Epoch: 12 Current loss: 2.255226; validation set accu: 0.205600	
Epoch: 13 Current loss: 2.250010; validation set accu: 0.214800	
Epoch: 14 Current loss: 2.244707; validation set accu: 0.225300	
Epoch: 15 Current loss: 2.239949; validation set accu: 0.234300	
Epoch: 16 Current loss: 2.234951; validation set accu: 0.243900	
Epoch: 17 Current loss: 2.230275; validation set accu: 0.251600	
Epoch: 18 Current loss: 2.225782; validation set accu: 0.260600	
Epoch: 19 Current loss: 2.221612; validation set accu: 0.268700	
Epoch: 20 Current loss: 2.217237; validation set accu: 0.275700

**the test code:**

require 'torch'
require 'nn'
require 'optim'
require 'cunn'
require 'cutorch'
local mnist = require 'mnist'

local fullset = mnist.traindataset()
local testset = mnist.testdataset()

local trainset = {
    size = 50000,
    data = fullset.data[{{1,50000}}]:double(),
    label = fullset.label[{{1,50000}}]
}
trainset.data = trainset.data - trainset.data:mean()
trainset.data = trainset.data:cuda()
trainset.label = trainset.label:cuda()

local validationset = {
    size = 10000,
    data = fullset.data[{{50001,60000}}]:double(),
    label = fullset.label[{{50001,60000}}]
}
validationset.data = validationset.data - validationset.data:mean()
validationset.data = validationset.data:cuda()
validationset.label = validationset.label:cuda()

local model = nn.Sequential()
model:add(nn.Reshape(1, 28, 28))
model:add(nn.MulConstant(1/256.0*3.2))
model:add(nn.SpatialConvolutionMM(1, 20, 5, 5, 1, 1, 0, 0))
model:add(nn.SpatialMaxPooling(2, 2 , 2, 2, 0, 0))
model:add(nn.SpatialConvolutionMM(20, 50, 5, 5, 1, 1, 0, 0))
model:add(nn.SpatialMaxPooling(2, 2 , 2, 2, 0, 0))
model:add(nn.Reshape(4*4*50))
model:add(nn.Linear(4*4*50, 500))
model:add(nn.ReLU())
model:add(nn.Linear(500, 10))
model:add(nn.LogSoftMax())

model = require('weight-init')(model, 'xavier')
model = model:cuda()

x, dl_dx = model:getParameters()

local criterion = nn.ClassNLLCriterion():cuda()

local sgd_params = {
   learningRate = 1e-2,
   learningRateDecay = 1e-4,
   weightDecay = 1e-3,
   momentum = 1e-4
}

local training = function(batchSize)
    local current_loss = 0
    local count = 0
    local shuffle = torch.randperm(trainset.size)
    batchSize = batchSize or 2000
    for t = 0, trainset.size-1, batchSize do
        -- setup inputs and targets for batch iteration
        local size = math.min(t + batchSize, trainset.size) - t
        local inputs = torch.Tensor(size, 28, 28):cuda()
        local targets = torch.Tensor(size):cuda()
        for i = 1, size do
            inputs[i] = trainset.data[shuffle[i+t]]
            targets[i] = trainset.label[shuffle[i+t]] + 1
        end

        local feval = function(x_new)
            local miniBatchSize = 200
            if x ~= x_new then x:copy(x_new) end   -- reset data
            dl_dx:zero()

            --[[ ------------------ method 1 original batch
            local outputs = model:forward(inputs)
            local loss = criterion:forward(outputs, targets)
            local gradInput = criterion:backward(outputs, targets)
            model:backward(inputs, gradInput)
            --]]

            --[[ ------------------ method 2 iter-size with batch
            local loss = 0
            for idx = 1, batchSize, miniBatchSize do
                local outputs = model:forward(inputs[{{idx, idx + miniBatchSize - 1}}])
                loss = loss + criterion:forward(outputs, targets[{{idx, idx + miniBatchSize - 1}}])
                local gradInput = criterion:backward(outputs, targets[{{idx, idx + miniBatchSize - 1}}])
                model:backward(inputs[{{idx, idx + miniBatchSize - 1}}], gradInput)
            end
            dl_dx:mul(1.0 * miniBatchSize / batchSize)
            loss = loss * miniBatchSize / batchSize
            --]]
            

            ---[[ ------------------ method 3 mini-batch in batch
            local outputs = torch.Tensor(batchSize, 10):zero():cuda()
            for idx = 1, batchSize, miniBatchSize do
                outputs[{{idx, idx + miniBatchSize - 1}}]:copy(model:forward(inputs[{{idx, idx + miniBatchSize - 1}}]))
            end
            local loss = criterion:forward(outputs, targets)
            local gradInput = criterion:backward(outputs, targets)            
            for idx = 1, batchSize, miniBatchSize do
                model:backward(inputs[{{idx, idx + miniBatchSize - 1}}], gradInput[{{idx, idx + miniBatchSize - 1}}])
            end
            dl_dx:mul(1.0 * miniBatchSize / batchSize)
            --]]

            return loss, dl_dx
        end

        _, fs = optim.sgd(feval, x, sgd_params)

        count = count + 1
        current_loss = current_loss + fs[1]
    end

    return current_loss / count   -- normalize loss
end

local eval = function(dataset, batchSize)
    local count = 0
    batchSize = batchSize or 200
    
    for i = 1, dataset.size, batchSize do
        local size = math.min(i + batchSize - 1, dataset.size) - i
        local inputs = dataset.data[{{i,i+size-1}}]:cuda()
        local targets = dataset.label[{{i,i+size-1}}]:long():cuda()
        local outputs = model:forward(inputs)
        local _, indices = torch.max(outputs, 2)
        indices:add(-1):cuda()
        local guessed_right = indices:eq(targets):sum()
        count = count + guessed_right
    end

    return count / dataset.size
end

local max_iters = 50
local last_accuracy = 0
local decreasing = 0
local threshold = 1 -- how many deacreasing epochs we allow
for i = 1,max_iters do
   -- timer = torch.Timer()
  
    model:training()
    local loss = training()
    
    model:evaluate()
    local accuracy = eval(validationset)
    print(string.format('Epoch: %d Current loss: %4f; validation set accu: %4f', i, loss, accuracy))
    if accuracy < last_accuracy then
        if decreasing > threshold then break end
        decreasing = decreasing + 1
    else
        decreasing = 0
    end
    last_accuracy = accuracy
    
    --print('    Time elapsed: ' .. i .. 'iter: ' .. timer:time().real .. ' seconds')
end

testset.data = testset.data:double()
eval(testset)

**weight-init.lua used in the test code**
--
-- Different weight initialization methods
--
-- > model = require('weight-init')(model, 'heuristic')
--
require("nn")

-- "Efficient backprop"
-- Yann Lecun, 1998
local function w_init_heuristic(fan_in, fan_out)
   return math.sqrt(1/(3*fan_in))
end

-- "Understanding the difficulty of training deep feedforward neural networks"
-- Xavier Glorot, 2010
local function w_init_xavier(fan_in, fan_out)
   return math.sqrt(2/(fan_in + fan_out))
end

-- "Understanding the difficulty of training deep feedforward neural networks"
-- Xavier Glorot, 2010
local function w_init_xavier_caffe(fan_in, fan_out)
   return math.sqrt(1/fan_in)
end

-- "Delving Deep into Rectifiers: Surpassing Human-Level Performance on ImageNet Classification"
-- Kaiming He, 2015
local function w_init_kaiming(fan_in, fan_out)
   return math.sqrt(4/(fan_in + fan_out))
end

local function w_init(net, arg)
   -- choose initialization method
   local method = nil
   if     arg == 'heuristic'    then method = w_init_heuristic
   elseif arg == 'xavier'       then method = w_init_xavier
   elseif arg == 'xavier_caffe' then method = w_init_xavier_caffe
   elseif arg == 'kaiming'      then method = w_init_kaiming
   else
      assert(false)
   end

   -- loop over all convolutional modules
   for i = 1, #net.modules do
      local m = net.modules[i]
      if m.__typename == 'nn.SpatialConvolution' then
         m:reset(method(m.nInputPlane*m.kH*m.kW, m.nOutputPlane*m.kH*m.kW))
      elseif m.__typename == 'nn.SpatialConvolutionMM' then
         m:reset(method(m.nInputPlane*m.kH*m.kW, m.nOutputPlane*m.kH*m.kW))
      elseif m.__typename == 'cudnn.SpatialConvolution' then
         m:reset(method(m.nInputPlane*m.kH*m.kW, m.nOutputPlane*m.kH*m.kW))
      elseif m.__typename == 'nn.LateralConvolution' then
         m:reset(method(m.nInputPlane*1*1, m.nOutputPlane*1*1))
      elseif m.__typename == 'nn.VerticalConvolution' then
         m:reset(method(1*m.kH*m.kW, 1*m.kH*m.kW))
      elseif m.__typename == 'nn.HorizontalConvolution' then
         m:reset(method(1*m.kH*m.kW, 1*m.kH*m.kW))
      elseif m.__typename == 'nn.Linear' then
         m:reset(method(m.weight:size(2), m.weight:size(1)))
      elseif m.__typename == 'nn.TemporalConvolution' then
         m:reset(method(m.weight:size(2), m.weight:size(1)))            
      end

      if m.bias then
         m.bias:zero()
      end
   end
   return net
end

return w_init

**Comment by soumith** [comment]
This is the place you want to post such topics: https://groups.google.com/forum/#!forum/torch7

**Comment by darkknightzh** [comment]
sorry， please delete this post.  thank you
